core/controllers/base_test.py

```diff
@@ -593,3 +595,64 @@ def test_get_items(self):
                 'param2=value%20with%20%26%20%2B%20-%20/&'
                 'param3=value%20with%20.%20%%20@%20123%20=%20!%20%3C%3E')
             self.assertDictContainsSubset(params, result)
+
+
+class ControllerClassNameTest(test_utils.GenericTestBase):
+
+    def test_controller_class_names(self):
+        """This function checks that all controller class names end with
+        either 'Handler' or 'Page'.
+        """
+        for url in main.URLS:
+            if not isinstance(url, tuple):
```

Comment:
Why this condition? Can we remove it?

Reply:
That's pretty non-obvious, and it seems to me like tuples are being appended to MAPREDUCE_HANDLERS anyway. So it is possible that this for loop is actually not testing anything. I think you want to include a count that increments by 1 for each URL that's being tested, and assert that that count is greater than 30 (or whatever). See test_every_method_has_decorator() in base_test.py for example, there's a check on the handler count at the end.

You should also have both: (a) a comment explaining the "why" behind this condition, and (b) a more robust check (perhaps isinstance of routes.RedirectRoute or something like that).

Reply:
I am sorry but I dont quite understand you. Why should we assert the count? Can't the number of controllers increase in future? And the condition is checking whether the URLs are redirect routes or not and I can go the other way around to actually test isinstance of routes.RedirectRoute as you suggested and some comments explaining the reason. Should I leave out the counting thing?

Reply:
If you assert that the count is > X, then that ensures that at least X handlers are being tested, and is a condition that will still hold even if we add new handlers. It's a safety check that I think is worth including to prevent errors in the test code itself, especially if the main codebase changes under it.

Yes, please test isinstance of routes.RedirectRoute and add some comments explaining the reason. That is more specific.

> But it will be easier to test isinstance of tuples so should I do according to the test_every_method_has_decorator() to include continue and all?

I don't understand this comment. Could you elaborate?

Reply:
@seanlip I made the changes.
> I don't understand this comment. Could you elaborate?

I just wanted to ask that should I use continue keyword whenever isinstance of tuple is found but that was not such a good idea after all.